kadi/events/__init__.py

```diff
@@ -42,22 +42,66 @@
 """
 
 import os
-import django
-from django.conf import settings
-from django.utils.functional import empty
-
-# If we are running standalone then this ENV var is not set.  Need to explicitly
-# do the setup and then import query module attributes for standalone.
-# For WSGI server the env var is set in wsgi.py.
-# For the dev server it is set in manage.py.
+import importlib
 
 # In addition, set DJANGO_ALLOW_ASYNC_UNSAFE, to avoid exception seen running in
 # Jupyter notebook: SynchronousOnlyOperation: You cannot call this from an async
 # context. See: https://stackoverflow.com/questions/59119396
 
-if settings._wrapped is empty:
-    if 'DJANGO_SETTINGS_MODULE' not in os.environ:
-        os.environ['DJANGO_SETTINGS_MODULE'] = 'kadi.settings'
-    os.environ["DJANGO_ALLOW_ASYNC_UNSAFE"] = "true"
-    django.setup()
-    from .query import *  # noqa
+os.environ['DJANGO_SETTINGS_MODULE'] = 'kadi.settings'
+os.environ["DJANGO_ALLOW_ASYNC_UNSAFE"] = "true"
+
+# Below is a list of every event query function + the `kadi.events.models`
+# module that is imported in `kadi.events.query`.
+#
+# There is a little subtlety here. See discussion in
+# https://github.com/sot/kadi/pull/231, but the upshot is that we need
+# `from kadi.events import models` to run `django.setup()`. This syntax is what
+# gets used in `update_events.py`. By including 'models' in `__all__` that
+# imports it via `query.py` and thus runs django setup.
+#
+# `manage.py` imports the same module but runs `django.setup()` before importing
+# it. However, it appears to import this via the equivalent of
+# `import kadi.events.models`, which does NOT go through `query.py`.
+# With this trick of lazy loading, both pathways work!
```

Comment:
but the upshot is that `django.setup()` must be called only once. Within kadi, `django.setup()` is called in `kadi.query`. Where the call to `django.setup()` actually happens depends on the application:
- `manage.py` calls `django.setup()` and then imports kadi.models (and it never imports `kadi.query`)
- `update_events.py` does `from kadi.events import models`, which calls `events. __getattr__`, which in turn imports `kadi.query`
- in any case, `kadi.query` should be imported before models can be used.